README.md
# About

Ansi utilities packages to handle string with ansi characters.

## Packages

| Project               | Status                                                       | Description                                             |
| --------------------- | ------------------------------------------------------------ | ------------------------------------------------------- |
| [ansi](packages/ansi/)          | [![ansi](https://img.shields.io/pub/v/ansi?color=blue)](https://pub.dev/packages/ansi)                   | Terminal string styling done right.                         |
| [ansi_codes](packages/ansi_codes/)          | [![ansi_codes](https://img.shields.io/pub/v/ansi_codes?color=blue)](https://pub.dev/packages/ansi_codes)                   | ANSI escape codes for styling strings in the terminal.                         |
| [ansi_regex](packages/ansi_regex/)          | [![ansi_regex](https://img.shields.io/pub/v/ansi_regex?color=blue)](https://pub.dev/packages/ansi_regex)                   | Regular expression for matching ANSI escape codes.                         |
| [ansi_slice](packages/ansi_slice/)          | [![ansi_slice](https://img.shields.io/pub/v/ansi_slice?color=blue)](https://pub.dev/packages/ansi_slice)                   | Slice a string with ANSI escape codes.                         |
| [ansi_strip](packages/ansi_strip/)          | [![ansi_strip](https://img.shields.io/pub/v/ansi_strip?color=blue)](https://pub.dev/packages/ansi_strip)                   | Strip ANSI escape codes from a string                         |
| [ansi_wrap](packages/ansi_wrap/)          | [![ansi_wrap](https://img.shields.io/pub/v/ansi_wrap?color=blue)](https://pub.dev/packages/ansi_wrap)                   | Wordwrap a string with ANSI escape codes.                         |



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## 2023-10-09

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`ansi` - `v0.4.1`](#ansi---v041)
 - [`ansi_slice` - `v0.1.1+1`](#ansi_slice---v0111)
 - [`ansi_wrap` - `v0.1.2+2`](#ansi_wrap---v0122)

---

#### `ansi` - `v0.4.1`

 - **FEAT**: make style apply to Object. ([0a657740](https://github.com/hyiso/ansi/commit/0a6577402c9ef46f4fa44404a3e1eb28dd9a87b8))

#### `ansi_slice` - `v0.1.1+1`

 - **DOCS**: change README. ([4856728b](https://github.com/hyiso/ansi/commit/4856728b579951ce4da7772ba0bab406c9341235))

#### `ansi_wrap` - `v0.1.2+2`

 - **DOCS**: change README. ([4856728b](https://github.com/hyiso/ansi/commit/4856728b579951ce4da7772ba0bab406c9341235))


## 2023-08-31

### Changes

---

Packages with breaking changes:

 - [`ansi` - `v0.4.0`](#ansi---v040)

Packages with other changes:

 - There are no other changes in this release.

---

#### `ansi` - `v0.4.0`

 - **BREAKING** **FEAT**: remove deprecated global instance `ansi`. ([ddc1068e](https://github.com/hyiso/ansi/commit/ddc1068e46c0c708a5b27808b145ab9d2b91fb51))
 - **BREAKING** **FEAT**: make `ansi` a pure package. ([9a6f6e08](https://github.com/hyiso/ansi/commit/9a6f6e0853cf4ded0710601215b500b1b204735f))


## 2023-08-25

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`ansi` - `v0.3.2`](#ansi---v032)

---

#### `ansi` - `v0.3.2`

 - **FEAT**: support running in test. ([8824d598](https://github.com/hyiso/ansi/commit/8824d598d07b1ae2e8bc26fc1d304862e5ea7111))


## 2023-08-25

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`ansi` - `v0.3.1`](#ansi---v031)

---

#### `ansi` - `v0.3.1`

 - **FEAT**: pub.dev score. ([afdb52dc](https://github.com/hyiso/ansi/commit/afdb52dcad6c26706a037d1ae62fade7fb6026c9))


## 2023-08-19

### Changes

---

Packages with breaking changes:

 - [`ansi` - `v0.3.0`](#ansi---v030)

Packages with other changes:

 - There are no other changes in this release.

---

#### `ansi` - `v0.3.0`

 - **BREAKING** **FEAT**: export extension `AnsiString` and style methods, mark `Ansi` and `ansi` as Deprecated. ([48681db1](https://github.com/hyiso/ansi/commit/48681db1bc747b85a83a6cc13b8b49bd3665b3ec))


## 2023-03-22

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`ansi_regex` - `v0.1.2`](#ansi_regex---v012)
 - [`ansi_strip` - `v0.1.1+1`](#ansi_strip---v0111)
 - [`ansi_wrap` - `v0.1.2+1`](#ansi_wrap---v0121)

Packages with dependency updates only:

> Packages listed below depend on other packages in this workspace that have had changes. Their versions have been incremented to bump the minimum dependency versions of the packages they depend upon in this project.

 - `ansi_strip` - `v0.1.1+1`
 - `ansi_wrap` - `v0.1.2+1`

---

#### `ansi_regex` - `v0.1.2`

 - **FEAT**: support `multiLine` parameter, expand regex and add test. ([dae5aae5](https://github.com/hyiso/ansi/commit/dae5aae505de4468db4252a66ec32d5964cc04b7))


## 2023-03-22

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`ansi` - `v0.2.2`](#ansi---v022)
 - [`ansi_codes` - `v0.1.1`](#ansi_codes---v011)
 - [`ansi_regex` - `v0.1.1`](#ansi_regex---v011)
 - [`ansi_slice` - `v0.1.1`](#ansi_slice---v011)
 - [`ansi_strip` - `v0.1.1`](#ansi_strip---v011)
 - [`ansi_wrap` - `v0.1.2`](#ansi_wrap---v012)

---

#### `ansi` - `v0.2.2`

 - **FEAT**: format codes. ([c56d1d4d](https://github.com/hyiso/ansi/commit/c56d1d4d08a658743ba799b142a0f21298cab6c3))

#### `ansi_codes` - `v0.1.1`

 - **FEAT**: format codes. ([c56d1d4d](https://github.com/hyiso/ansi/commit/c56d1d4d08a658743ba799b142a0f21298cab6c3))

#### `ansi_regex` - `v0.1.1`

 - **FEAT**: format codes. ([c56d1d4d](https://github.com/hyiso/ansi/commit/c56d1d4d08a658743ba799b142a0f21298cab6c3))

#### `ansi_slice` - `v0.1.1`

 - **FEAT**: format codes. ([c56d1d4d](https://github.com/hyiso/ansi/commit/c56d1d4d08a658743ba799b142a0f21298cab6c3))

#### `ansi_strip` - `v0.1.1`

 - **FEAT**: format codes. ([c56d1d4d](https://github.com/hyiso/ansi/commit/c56d1d4d08a658743ba799b142a0f21298cab6c3))

#### `ansi_wrap` - `v0.1.2`

 - **FEAT**: format codes. ([c56d1d4d](https://github.com/hyiso/ansi/commit/c56d1d4d08a658743ba799b142a0f21298cab6c3))


## 2023-03-19

### Changes

---

Packages with breaking changes:

 - [`ansi` - `v0.2.0`](#ansi---v020)

Packages with other changes:

 - There are no other changes in this release.

---

#### `ansi` - `v0.2.0`

 - **BREAKING** **FEAT**(ansi): depends on `ansi_codes`. ([c49cf0ec](https://github.com/hyiso/ansi/commit/c49cf0ecbc576eff4ec84c704df4325286930495))


## 2023-03-19

### Changes

---

Packages with breaking changes:

 - There are no breaking changes in this release.

Packages with other changes:

 - [`ansi` - `v0.1.0+1`](#ansi---v0101)

---

#### `ansi` - `v0.1.0+1`

 - **DOCS**(ansi): update README.md. ([bfcd6ac6](https://github.com/hyiso/ansi/commit/bfcd6ac6a760ed0866f425fe6e2272d1c12daaf5))



test/ansi_test.dart
import 'package:test/test.dart';
import 'package:ansi/ansi.dart';

void main() {
  group('Ansi', () {
    test('should reset text', () {
      expect(Ansi.reset(), '\x1B[0m');
    });

    test('should make text bold', () {
      expect(Ansi.bold('text'), '\x1B[1mtext\x1B[0m');
    });

    test('should make text dim', () {
      expect(Ansi.dim('text'), '\x1B[2mtext\x1B[0m');
    });

    test('should make text italic', () {
      expect(Ansi.italic('text'), '\x1B[3mtext\x1B[0m');
    });

    test('should underline text', () {
      expect(Ansi.underline('text'), '\x1B[4mtext\x1B[0m');
    });

    test('should inverse text', () {
      expect(Ansi.inverse('text'), '\x1B[7mtext\x1B[0m');
    });

    test('should color text black', () {
      expect(Ansi.black('text'), '\x1B[30mtext\x1B[0m');
    });

    test('should color text red', () {
      expect(Ansi.red('text'), '\x1B[31mtext\x1B[0m');
    });

    test('should color text green', () {
      expect(Ansi.green('text'), '\x1B[32mtext\x1B[0m');
    });

    test('should color text yellow', () {
      expect(Ansi.yellow('text'), '\x1B[33mtext\x1B[0m');
    });

    test('should color text blue', () {
      expect(Ansi.blue('text'), '\x1B[34mtext\x1B[0m');
    });

    test('should color text magenta', () {
      expect(Ansi.magenta('text'), '\x1B[35mtext\x1B[0m');
    });

    test('should color text cyan', () {
      expect(Ansi.cyan('text'), '\x1B[36mtext\x1B[0m');
    });

    test('should color text white', () {
      expect(Ansi.white('text'), '\x1B[37mtext\x1B[0m');
    });

    test('should color background black', () {
      expect(Ansi.bgBlack('text'), '\x1B[40mtext\x1B[0m');
    });

    test('should color background red', () {
      expect(Ansi.bgRed('text'), '\x1B[41mtext\x1B[0m');
    });

    test('should color background green', () {
      expect(Ansi.bgGreen('text'), '\x1B[42mtext\x1B[0m');
    });

    test('should color background yellow', () {
      expect(Ansi.bgYellow('text'), '\x1B[43mtext\x1B[0m');
    });

    test('should color background blue', () {
      expect(Ansi.bgBlue('text'), '\x1B[44mtext\x1B[0m');
    });

    test('should color background magenta', () {
      expect(Ansi.bgMagenta('text'), '\x1B[45mtext\x1B[0m');
    });

    test('should color background cyan', () {
      expect(Ansi.bgCyan('text'), '\x1B[46mtext\x1B[0m');
    });

    test('should color background white', () {
      expect(Ansi.bgWhite('text'), '\x1B[47mtext\x1B[0m');
    });
  });

  group('sliceAnsi', () {
    test('should return the original string if no ANSI codes are used', () {
      expect(sliceAnsi('Hello world'), equals('Hello world'));
    });

    test('should remove ANSI color codes from a string', () {
      expect(
          sliceAnsi('\u001b[31mHello\u001b[32m \u001b[33mworld\u001b[0m'),
          equals('Hello world'));
    });
    
    test('should remove all ANSI escape sequences from a string', () {
      expect(
          sliceAnsi('\u001b[31mHello\u001b[1;32m \u001b[33mworld\u001b[0m\n'),
          equals('Hello world\n'));
    });
  });

  group('stripAnsi', () {
    test('should strip ANSI escape codes from a string', () {
      final formattedString = '\x1B[31mError:\x1B[0m \x1B[33mSomething went wrong\x1B[0m';
      final expectedString = 'Error: Something went wrong';
      final strippedString = stripAnsi(formattedString);

      expect(strippedString, equals(expectedString));
    });

    test('should return the same string when no ANSI escape codes are present', () {
      final plainString = 'Hello, world!';
      final strippedString = stripAnsi(plainString);

      expect(strippedString, equals(plainString));
    });
  });

}


test/ansi_slice_test.dart
import 'dart:convert';

import 'package:ansi/ansi.dart';
import 'package:ansi/src/slice.dart';
import 'package:ansi_styles/ansi_styles.dart';
import 'package:test/test.dart';

void main() {
  final fixture = AnsiStyles.red('the ') + AnsiStyles.green('quick ') + AnsiStyles.blue('brown ') + AnsiStyles.cyan('fox ') + AnsiStyles.yellow('jumped ');
  final stripped = stripAnsi(fixture);
  test('main', () {
	// The slice should behave exactly as a regular JS slice behaves
	for (var index = 0; index < 20; index++) {
		for (var index2 = 19; index2 > index; index2--) {
			var nativeSlice = stripped.substring(index, index2);
			var ansiSlice = sliceAnsi(fixture, index, index2);
			expect(nativeSlice, stripAnsi(ansiSlice));
		}
	}

	var a = jsonEncode('\u001B[31mthe \u001B[39m\u001B[32mquick \u001B[39m');
	var b = jsonEncode('\u001B[34mbrown \u001B[39m\u001B[36mfox \u001B[39m');
	var c = jsonEncode('\u001B[31m \u001B[39m\u001B[32mquick \u001B[39m\u001B[34mbrown \u001B[39m\u001B[36mfox \u001B[39m');

	expect(jsonEncode(sliceAnsi(fixture, 0, 10)), a);
	expect(jsonEncode(sliceAnsi(fixture, 10, 20)), b);
	expect(jsonEncode(sliceAnsi(fixture, 3, 20)), c);
});

test('supports fullwidth characters', () {
	expect(sliceAnsi('안녕하세', 0, 4), '안녕');
});

test('supports unicode surrogate pairs', () {
	expect(sliceAnsi('a\uD83C\uDE00BC', 0, 2), 'a\uD83C\uDE00');
});

test('doesn\'t add unnecessary escape codes', () {
	expect(sliceAnsi('\u001B[31municorn\u001B[39m', 0, 3), '\u001B[31muni\u001B[39m');
});

test('can slice a normal character before a colored character', () {
	expect(sliceAnsi('a\u001B[31mb\u001B[39m', 0, 1), 'a');
});

test('can slice a normal character after a colored character', () {
	expect(sliceAnsi('\u001B[31ma\u001B[39mb', 1, 2), 'b');
});

// See https://github.com/AnsiStyles/slice-ansi/issues/22
test('can slice a string styled with both background and foreground', () {
	// Test string: 'AnsiStyles.bgGreen.black('test');'
	expect(sliceAnsi('\u001B[42m\u001B[30mtest\u001B[39m\u001B[49m', 0, 1), '\u001B[42m\u001B[30mt\u001B[39m\u001B[49m');
});

test('can slice a string styled with modifier', () {
	// Test string: 'AnsiStyles.underline('test');'
	expect(sliceAnsi('\u001B[4mtest\u001B[24m', 0, 1), '\u001B[4mt\u001B[24m');
});

test('can slice a string with unknown ANSI color', () {
	expect(sliceAnsi('\u001B[20mTEST\u001B[49m', 0, 4), '\u001B[20mTEST\u001B[0m');
	expect(sliceAnsi('\u001B[1001mTEST\u001B[49m', 0, 3), '\u001B[1001mTES\u001B[0m');
	expect(sliceAnsi('\u001B[1001mTEST\u001B[49m', 0, 2), '\u001B[1001mTE\u001B[0m');
});

test('weird null issue', () {
	var s = '\u001B[1mautotune.flipCoin("easy as") ? 🎂 : 🍰 \u001B[33m★\u001B[39m\u001B[22m';
	var result = sliceAnsi(s, 38);
	expect(result.contains('null'), false);
});

test('support true color escape sequences', () {
	expect(sliceAnsi('\u001B[1m\u001B[48;2;255;255;255m\u001B[38;2;255;0;0municorn\u001B[39m\u001B[49m\u001B[22m', 0, 3), '\u001B[1m\u001B[48;2;255;255;255m\u001B[38;2;255;0;0muni\u001B[22m\u001B[49m\u001B[39m');
});

// See https://github.com/AnsiStyles/slice-ansi/issues/24
test('doesn\'t add extra escapes', () {
	var output = '${AnsiStyles.black.bgYellow(' RUNS ')}  ${AnsiStyles.green('test')}';
	expect(sliceAnsi(output, 0, 7), '${AnsiStyles.black.bgYellow(' RUNS ')} ');
	expect(sliceAnsi(output, 0, 8), '${AnsiStyles.black.bgYellow(' RUNS ')}  ');
	expect(jsonEncode(sliceAnsi('\u001B[31m' + output, 0, 4)),jsonEncode('\u001B[31m${AnsiStyles.black.bgYellow(' RUN')}'));
});

// See https://github.com/AnsiStyles/slice-ansi/issues/26
test('does not lose fullwidth characters', () {
	expect(sliceAnsi('古古test', 0), '古古test');
});

test('slice links', () {
	var link = '\u001B]8;;https://google.com\u0007Google\u001B]8;;\u0007';
	expect(sliceAnsi(link, 0, 6), link);
});
}

lib/ansi.dart
import 'dart:io';

import 'dart:math';

const _kEscape = '\x1B[';
class Ansi {

  static String reset() => _code(0);
  static String bold(String text) => '${_code(1)}$text${reset()}';
  static String dim(String text) => '${_code(2)}$text${reset()}';
  static String italic(String text) => '${_code(3)}$text${reset()}';
  static String underline(String text) => '${_code(4)}$text${reset()}';
  static String inverse(String text) => '${_code(7)}$text${reset()}';

  static String black(String text) => '${_code(30)}$text${reset()}';
  static String red(String text) => '${_code(31)}$text${reset()}';
  static String green(String text) => '${_code(32)}$text${reset()}';
  static String yellow(String text) => '${_code(33)}$text${reset()}';
  static String blue(String text) => '${_code(34)}$text${reset()}';
  static String magenta(String text) => '${_code(35)}$text${reset()}';
  static String cyan(String text) => '${_code(36)}$text${reset()}';
  static String white(String text) => '${_code(37)}$text${reset()}';

  static String bgBlack(String text) => '${_code(40)}$text${reset()}';
  static String bgRed(String text) => '${_code(41)}$text${reset()}';
  static String bgGreen(String text) => '${_code(42)}$text${reset()}';
  static String bgYellow(String text) => '${_code(43)}$text${reset()}';
  static String bgBlue(String text) => '${_code(44)}$text${reset()}';
  static String bgMagenta(String text) => '${_code(45)}$text${reset()}';
  static String bgCyan(String text) => '${_code(46)}$text${reset()}';
  static String bgWhite(String text) => '${_code(47)}$text${reset()}';

  static String _code(int code) => stdout.supportsAnsiEscapes && stdioType(stdout) == StdioType.terminal ? '$_kEscape${code}m' : '';
}

String sliceAnsi(String text, {int start = 0, int? end}) {
  end ??= text.length;
  final pattern = RegExp('\x1B\\[[0-9;]*m');
  final buffer = StringBuffer();
  var skip = 0;
  for (var i = 0; i < text.length; i++) {
    final char = text[i];
    if (char == '\x1B') {
      final match = pattern.matchAsPrefix(text.substring(i));
      if (match != null) {
        i += match.end - match.start - 1;
        continue;
      }
    }
    if (skip > 0) {
      skip--;
      continue;
    }
    if (buffer.length >= end) {
      break;
    }
    if (buffer.length >= start) {
      buffer.write(char);
    }
  }
  return buffer.toString();
}

String stripAnsi(String text) {
  return text.replaceAll(RegExp('\x1B\\[[0-9;]+m', multiLine: true), '');
}


lib/src/slice.dart

import 'dart:math';

import 'package:ansi/src/codes.dart';

import 'utils.dart';

final astralRegex = RegExp(r'^[\uD800-\uDBFF][\uDC00-\uDFFF]$');

final kEscapes = [
	'\u001B',
	'\u009B'
];

String wrapAnsi(dynamic code) => '${kEscapes[0]}[${code}m';

String checkAnsi(List<String> ansiCodes, [bool isEscapes = false, String? endAnsiCode]) {
	var output = <String>[];
  var removes = <String>[];

	for (var ansiCode in ansiCodes) {
		final ansiCodeOrigin = ansiCode;
		if (ansiCode.contains(';')) {
			ansiCode = '${ansiCode.split(';')[0][0]}0';
		}

		final item = codes[int.tryParse(ansiCode, radix: 10)];
		if (item != null) {
			final indexEscape = ansiCodes.indexOf(item.toString());
			if (indexEscape == -1) {
				output.add(wrapAnsi(isEscapes ? item : ansiCodeOrigin));
			} else {
        removes.add(item.toString());
			}
		} else if (isEscapes) {
			output.add(wrapAnsi(0));
			break;
		} else {
			output.add(wrapAnsi(ansiCodeOrigin));
		}
	}
  ansiCodes.removeWhere((element) => removes.contains(element));

	if (isEscapes) {
		output = output.toSet().toList();

		if (endAnsiCode != null) {
			final fistEscapeCode = wrapAnsi(codes[int.tryParse(endAnsiCode, radix: 10)]);
			output = output.fold([], (current, next) => next == fistEscapeCode ? [next, ...current] : [...current, next]);
		}
	}

	return output.join('');
}

/// Slice a string with ANSI escape codes
String sliceAnsi(String input, int begin, [int? end]) {
	final ansiCodes = <String>[];

	int stringEnd = end ?? input.length;
	bool isInsideEscape = false;
	String? ansiCode;
	int visible = 0;
	String output = '';

	for (var i = 0; i < input.length; i++) {
		bool leftEscape = false;
    final character = input[i];

		if (kEscapes.contains(character)) {
			final code = RegExp(r'\d[^m]*').firstMatch(input.substring(i, min(i + 18, input.length)));
			ansiCode = code?.group(0);

			if (visible < stringEnd) {
				isInsideEscape = true;

				if (ansiCode != null) {
					ansiCodes.add(ansiCode);
				}
			}
		} else if (isInsideEscape && character == 'm') {
			isInsideEscape = false;
			leftEscape = true;
		}

		if (!isInsideEscape && !leftEscape) {
			visible++;
		}

		if (!astralRegex.hasMatch(character) && isFullwidthCodePoint(character.codeUnitAt(0))) {
			visible++;

			if (end == null) {
				stringEnd++;
			}
		}

		if (visible > begin && visible <= stringEnd) {
			output += character;
		} else if (visible == begin && !isInsideEscape && ansiCode != null) {
			output = checkAnsi(ansiCodes);
		} else if (visible >= stringEnd) {
			output += checkAnsi(ansiCodes, true, ansiCode);
			break;
		}
	}

	return output;
}

lib/src/utils.dart
/// Check if the character represented by a given Unicode code point is fullwidth
bool isFullwidthCodePoint(int codePoint) {

	// Code points are derived from:
	// https://unicode.org/Public/UNIDATA/EastAsianWidth.txt
	return codePoint >= 0x1100 && (
		codePoint <= 0x115F || // Hangul Jamo
		codePoint == 0x2329 || // LEFT-POINTING ANGLE BRACKET
		codePoint == 0x232A || // RIGHT-POINTING ANGLE BRACKET
		// CJK Radicals Supplement .. Enclosed CJK Letters and Months
		(0x2E80 <= codePoint && codePoint <= 0x3247 && codePoint != 0x303F) ||
		// Enclosed CJK Letters and Months .. CJK Unified Ideographs Extension A
		(0x3250 <= codePoint && codePoint <= 0x4DBF) ||
		// CJK Unified Ideographs .. Yi Radicals
		(0x4E00 <= codePoint && codePoint <= 0xA4C6) ||
		// Hangul Jamo Extended-A
		(0xA960 <= codePoint && codePoint <= 0xA97C) ||
		// Hangul Syllables
		(0xAC00 <= codePoint && codePoint <= 0xD7A3) ||
		// CJK Compatibility Ideographs
		(0xF900 <= codePoint && codePoint <= 0xFAFF) ||
		// Vertical Forms
		(0xFE10 <= codePoint && codePoint <= 0xFE19) ||
		// CJK Compatibility Forms .. Small Form Variants
		(0xFE30 <= codePoint && codePoint <= 0xFE6B) ||
		// Halfwidth and Fullwidth Forms
		(0xFF01 <= codePoint && codePoint <= 0xFF60) ||
		(0xFFE0 <= codePoint && codePoint <= 0xFFE6) ||
		// Kana Supplement
		(0x1B000 <= codePoint && codePoint <= 0x1B001) ||
		// Enclosed Ideographic Supplement
		(0x1F200 <= codePoint && codePoint <= 0x1F251) ||
		// CJK Unified Ideographs Extension B .. Tertiary Ideographic Plane
		(0x20000 <= codePoint && codePoint <= 0x3FFFD)
	);
}


lib/src/codes.dart
const _styles = {
	'modifier': {
		'reset': [0, 0],
		// 21 isn't widely supported and 22 does the same thing
		'bold': [1, 22],
		'dim': [2, 22],
		'italic': [3, 23],
		'underline': [4, 24],
		'overline': [53, 55],
		'inverse': [7, 27],
		'hidden': [8, 28],
		'strikethrough': [9, 29],
	},
	'color': {
		'black': [30, 39],
		'red': [31, 39],
		'green': [32, 39],
		'yellow': [33, 39],
		'blue': [34, 39],
		'magenta': [35, 39],
		'cyan': [36, 39],
		'white': [37, 39],

		// Bright color
		'blackBright': [90, 39],
		'gray': [90, 39], // Alias of 'blackBright'
		'grey': [90, 39], // Alias of 'blackBright'
		'redBright': [91, 39],
		'greenBright': [92, 39],
		'yellowBright': [93, 39],
		'blueBright': [94, 39],
		'magentaBright': [95, 39],
		'cyanBright': [96, 39],
		'whiteBright': [97, 39],
	},
	'bgColor': {
		'bgBlack': [40, 49],
		'bgRed': [41, 49],
		'bgGreen': [42, 49],
		'bgYellow': [43, 49],
		'bgBlue': [44, 49],
		'bgMagenta': [45, 49],
		'bgCyan': [46, 49],
		'bgWhite': [47, 49],

		// Bright color
		'bgBlackBright': [100, 49],
		'bgGray': [100, 49], // Alias of 'bgBlackBright'
		'bgGrey': [100, 49], // Alias of 'bgBlackBright'
		'bgRedBright': [101, 49],
		'bgGreenBright': [102, 49],
		'bgYellowBright': [103, 49],
		'bgBlueBright': [104, 49],
		'bgMagentaBright': [105, 49],
		'bgCyanBright': [106, 49],
		'bgWhiteBright': [107, 49],
	},
};

Map<int, int> styleCodes() {
	final codes = <int, int>{};

	for (var group in _styles.values) {
		for (var style in group.values) {
			codes[style.first] = style.last;
		}
  }
	return codes;
}

final codes = styleCodes();